Allow topics to be deleted

yichenshen opened the pull request:
Deleting topics with posts that have votes previously causes a foreign
key error.

Because we originally expect posts to be reparented, children are not destroyed by association. As such, it has to be done manually.

weiqingtoh commented:
Seems better to make it slightly more explicit 👍 

Could you add model specs to test that the `before_destroy` callbacks are invoked correctly, or is that already done?

yichenshen commented:
Nope, they weren't there, but I've added them.

There was also a bug with the original destroy code. If a post has 3 levels of children, the first parent will call the normal destroy on its children, triggering reparenting instead of destruction. We still get a foreign key error since these children posts references topic.

I changed it to just flatten all the children when it's destroyed by association. That way the children get destroyed by the topic without error too.

weiqingtoh commented on spec/models/course/discussion/topic_spec.rb:
```diff
@@ -155,5 +155,42 @@
+
+        context 'when the posts have votes' do
+          before(:each) do
+            post.cast_vote!(user, 1)
```
Instead of explicitly calling `#cast_vote!` here, we can add traits to the `posts` factory (eg. `post_count`).

weiqingtoh commented on app/models/course/discussion/post.rb:
```diff
@@ -98,4 +99,12 @@ def parent_topic_consistency
+  # Should be called only when destroyed by association.
+  #
+  # We unset the children's parent id so they don't trigger a foreign key exception when the
+  # parent is marked for destruction first. They will be destroyed by association later.
```
Possible to add 1 more assumption here, that when `destroyed_by_association` is `true`, we assume that the entire `topic` will be deleted?

weiqingtoh commented:
I think once these are fixed, LGTM 👍

yichenshen commented:
I've added the comment for the assumption.

I wasn't too certain about the transient attribute part, but I added `vote_count` to posts and shifted the `cast_vote!` there. Was that what you meant? Thanks!

weiqingtoh commented:
Yup this is what I meant - this makes the creation of objects cleaner 👍

yichenshen commented:
Alright, thanks! I rebased to fix the broken specs.

weiqingtoh commented on spec/factories/course_discussion_posts.rb:
```diff
@@ -1,10 +1,22 @@
+    after(:create) do |post, evaluator|
+      next if evaluator.user.blank? || evaluator.vote_count == 0
+      evaluator.vote_count.times do
+        post.cast_vote!(evaluator.user, 1)
```
If `vote_count` is 1, and `user` is `nil`, will that be an issue? Might need to enumerate through the `user` attribute. 

Also, I read through the `#cast_vote` method - if we call the same method twice with the same user, it actually only creates 1 upvote. 

I'm thinking if there are other ways to define this transient attribute - possibly having `upvoters` and `downvoters` instead, each taking in an array of users - what do you think? Or if you have other ideas, you can suggest too!

yichenshen replied:
You're right, it'll only create a single vote, which wasn't what I intended it to be.

I thought of using a Hash like structure at first, but the `upvoters` and `downvoters` approach was cleaner. I've updated it accordingly.

weiqingtoh commented:
👍 thanks @yichenshen
